File: skills/wechatpay-basic-payment/references/1-商户/示例代码/Go/16-分账/QueryOrder.go
package main

import (
	"demo/wxpay_utility" // 引用微信支付工具库，参考 https://pay.weixin.qq.com/doc/v3/merchant/4015119334
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"
)

func main() {
	// TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/merchant/4013070756
	config, err := wxpay_utility.CreateMchConfig(
		"YOUR_MCHID",                 // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/merchant/4013070756
		"YOUR_CERT_SERIAL_NO",      // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013053053
		"/path/to/apiclient_key.pem", // 商户API证书私钥文件路径，本地文件路径
		"YOUR_PUB_KEY_ID",   // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013038816
		"/path/to/wxp_pub.pem",       // 微信支付公钥文件路径，本地文件路径
	)
	if err != nil {
		fmt.Println(err)
		return
	}

	request := &QueryOrderRequest{
		TransactionId: wxpay_utility.String("4208450740201411110007820472"),
		OutOrderNo:    wxpay_utility.String("P20150806125346"),
	}

	response, err := QueryOrder(config, request)
	if err != nil {
		fmt.Printf("请求失败: %+v\n", err)
		// TODO: 请求失败，根据状态码执行不同的处理
		return
	}

	// TODO: 请求成功，继续业务逻辑
	fmt.Printf("请求成功: %+v\n", response)
}

func QueryOrder(config *wxpay_utility.MchConfig, request *QueryOrderRequest) (response *OrdersEntity, err error) {
	const (
		host   = "https://api.mch.weixin.qq.com"
		method = "GET"
		path   = "/v3/profitsharing/orders/{out_order_no}"
	)

	reqUrl, err := url.Parse(fmt.Sprintf("%s%s", host, path))
	if err != nil {
		return nil, err
	}
	reqUrl.Path = strings.Replace(reqUrl.Path, "{out_order_no}", url.PathEscape(*request.OutOrderNo), -1)
	query := reqUrl.Query()
	if request.TransactionId != nil {
		query.Add("transaction_id", *request.TransactionId)
	}
	reqUrl.RawQuery = query.Encode()
	httpRequest, err := http.NewRequest(method, reqUrl.String(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Accept", "application/json")
	httpRequest.Header.Set("Wechatpay-Serial", config.WechatPayPublicKeyId())
	authorization, err := wxpay_utility.BuildAuthorization(config.MchId(), config.CertificateSerialNo(), config.PrivateKey(), method, reqUrl.RequestURI(), nil)
	if err != nil {
		return nil, err
	}
	httpRequest.Header.Set("Authorization", authorization)

	client := &http.Client{}
	httpResponse, err := client.Do(httpRequest)
	if err != nil {
		return nil, err
	}
	respBody, err := wxpay_utility.ExtractResponseBody(httpResponse)
	if err != nil {
		return nil, err
	}
	if httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300 {
		// 2XX 成功，验证应答签名
		err = wxpay_utility.ValidateResponse(
			config.WechatPayPublicKeyId(),
			config.WechatPayPublicKey(),
			&httpResponse.Header,
			respBody,
		)
		if err != nil {
			return nil, err
		}
		response := &OrdersEntity{}
		if err := json.Unmarshal(respBody, response); err != nil {
			return nil, err
		}

		return response, nil
	} else {
		return nil, wxpay_utility.NewApiException(
			httpResponse.StatusCode,
			httpResponse.Header,
			respBody,
		)
	}
}

type QueryOrderRequest struct {
	TransactionId *string `json:"transaction_id,omitempty"`
	OutOrderNo    *string `json:"out_order_no,omitempty"`
}

func (o *QueryOrderRequest) MarshalJSON() ([]byte, error) {
	type Alias QueryOrderRequest
	a := &struct {
		TransactionId *string `json:"transaction_id,omitempty"`
		OutOrderNo    *string `json:"out_order_no,omitempty"`
		*Alias
	}{
		// 序列化时移除非 Body 字段
		TransactionId: nil,
		OutOrderNo:    nil,
		Alias:         (*Alias)(o),
	}
	return json.Marshal(a)
}

type OrdersEntity struct {
	TransactionId *string               `json:"transaction_id,omitempty"`
	OutOrderNo    *string               `json:"out_order_no,omitempty"`
	OrderId       *string               `json:"order_id,omitempty"`
	State         *OrderStatus          `json:"state,omitempty"`
	Receivers     []OrderReceiverDetail `json:"receivers,omitempty"`
}

type OrderStatus string

func (e OrderStatus) Ptr() *OrderStatus {
	return &e
}

const (
	ORDERSTATUS_PROCESSING OrderStatus = "PROCESSING"
	ORDERSTATUS_FINISHED   OrderStatus = "FINISHED"
)

type OrderReceiverDetail struct {
	Amount      *int64            `json:"amount,omitempty"`
	Description *string           `json:"description,omitempty"`
	Type        *ReceiverType     `json:"type,omitempty"`
	Account     *string           `json:"account,omitempty"`
	Result      *DetailStatus     `json:"result,omitempty"`
	FailReason  *DetailFailReason `json:"fail_reason,omitempty"`
	CreateTime  *time.Time        `json:"create_time,omitempty"`
	FinishTime  *time.Time        `json:"finish_time,omitempty"`
	DetailId    *string           `json:"detail_id,omitempty"`
}

type ReceiverType string

func (e ReceiverType) Ptr() *ReceiverType {
	return &e
}

const (
	RECEIVERTYPE_MERCHANT_ID     ReceiverType = "MERCHANT_ID"
	RECEIVERTYPE_PERSONAL_OPENID ReceiverType = "PERSONAL_OPENID"
)

type DetailStatus string

func (e DetailStatus) Ptr() *DetailStatus {
	return &e
}

const (
	DETAILSTATUS_PENDING DetailStatus = "PENDING"
	DETAILSTATUS_SUCCESS DetailStatus = "SUCCESS"
	DETAILSTATUS_CLOSED  DetailStatus = "CLOSED"
)

type DetailFailReason string

func (e DetailFailReason) Ptr() *DetailFailReason {
	return &e
}

const (
	DETAILFAILREASON_ACCOUNT_ABNORMAL                DetailFailReason = "ACCOUNT_ABNORMAL"
	DETAILFAILREASON_NO_RELATION                     DetailFailReason = "NO_RELATION"
	DETAILFAILREASON_RECEIVER_HIGH_RISK              DetailFailReason = "RECEIVER_HIGH_RISK"
	DETAILFAILREASON_RECEIVER_REAL_NAME_NOT_VERIFIED DetailFailReason = "RECEIVER_REAL_NAME_NOT_VERIFIED"
	DETAILFAILREASON_NO_AUTH                         DetailFailReason = "NO_AUTH"
	DETAILFAILREASON_RECEIVER_RECEIPT_LIMIT          DetailFailReason = "RECEIVER_RECEIPT_LIMIT"
	DETAILFAILREASON_PAYER_ACCOUNT_ABNORMAL          DetailFailReason = "PAYER_ACCOUNT_ABNORMAL"
	DETAILFAILREASON_INVALID_REQUEST                 DetailFailReason = "INVALID_REQUEST"
)

请求分账回退

File: skills/wechatpay-basic-payment/references/1-商户/示例代码/Java/16-分账/QueryOrder.java
package com.java.demo;

import com.java.utils.WXPayUtility; // 引用微信支付工具库，参考：https://pay.weixin.qq.com/doc/v3/merchant/4014931831

import com.google.gson.annotations.SerializedName;
import com.google.gson.annotations.Expose;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 查询分账结果
 */
public class QueryOrder {
  private static String HOST = "https://api.mch.weixin.qq.com";
  private static String METHOD = "GET";
  private static String PATH = "/v3/profitsharing/orders/{out_order_no}";

  public static void main(String[] args) {
    // TODO: 请准备商户开发必要参数，参考：https://pay.weixin.qq.com/doc/v3/merchant/4013070756
    QueryOrder client = new QueryOrder(
      "YOUR_MCHID",                    // 商户号，是由微信支付系统生成并分配给每个商户的唯一标识符，商户号获取方式参考 https://pay.weixin.qq.com/doc/v3/merchant/4013070756
      "YOUR_CERT_SERIAL_NO",         // 商户API证书序列号，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013053053
      "/path/to/apiclient_key.pem",     // 商户API证书私钥文件路径，本地文件路径
      "YOUR_PUB_KEY_ID",      // 微信支付公钥ID，如何获取请参考 https://pay.weixin.qq.com/doc/v3/merchant/4013038816
      "/path/to/wxp_pub.pem"           // 微信支付公钥文件路径，本地文件路径
    );

    QueryOrderRequest request = new QueryOrderRequest();
    request.outOrderNo = "P20150806125346";
    request.transactionId = "4208450740201411110007820472";
    try {
      OrdersEntity response = client.run(request);
        // TODO: 请求成功，继续业务逻辑
        System.out.println(response);
    } catch (WXPayUtility.ApiException e) {
        // TODO: 请求失败，根据状态码执行不同的逻辑
        e.printStackTrace();
    }
  }

  public OrdersEntity run(QueryOrderRequest request) {
    String uri = PATH;
    uri = uri.replace("{out_order_no}", WXPayUtility.urlEncode(request.outOrderNo));
    Map<String, Object> args = new HashMap<>();
    args.put("transaction_id", request.transactionId);
    String queryString = WXPayUtility.urlEncode(args);
    if (!queryString.isEmpty()) {
        uri = uri + "?" + queryString;
    }

    Request.Builder reqBuilder = new Request.Builder().url(HOST + uri);
    reqBuilder.addHeader("Accept", "application/json");
    reqBuilder.addHeader("Wechatpay-Serial", wechatPayPublicKeyId);
    reqBuilder.addHeader("Authorization", WXPayUtility.buildAuthorization(mchid, certificateSerialNo, privateKey, METHOD, uri, null));
    reqBuilder.method(METHOD, null);
    Request httpRequest = reqBuilder.build();

    // 发送HTTP请求
    OkHttpClient client = new OkHttpClient.Builder().build();
    try (Response httpResponse = client.newCall(httpRequest).execute()) {
      String respBody = WXPayUtility.extractBody(httpResponse);
      if (httpResponse.code() >= 200 && httpResponse.code() < 300) {
        // 2XX 成功，验证应答签名
        WXPayUtility.validateResponse(this.wechatPayPublicKeyId, this.wechatPayPublicKey,
            httpResponse.headers(), respBody);

        // 从HTTP应答报文构建返回数据
        return WXPayUtility.fromJson(respBody, OrdersEntity.class);
      } else {
        throw new WXPayUtility.ApiException(httpResponse.code(), respBody, httpResponse.headers());
      }
    } catch (IOException e) {
      throw new UncheckedIOException("Sending request to " + uri + " failed.", e);
    }
  }

  private final String mchid;
  private final String certificateSerialNo;
  private final PrivateKey privateKey;
  private final String wechatPayPublicKeyId;
  private final PublicKey wechatPayPublicKey;

  public QueryOrder(String mchid, String certificateSerialNo, String privateKeyFilePath, String wechatPayPublicKeyId, String wechatPayPublicKeyFilePath) {
    this.mchid = mchid;
    this.certificateSerialNo = certificateSerialNo;
    this.privateKey = WXPayUtility.loadPrivateKeyFromPath(privateKeyFilePath);
    this.wechatPayPublicKeyId = wechatPayPublicKeyId;
    this.wechatPayPublicKey = WXPayUtility.loadPublicKeyFromPath(wechatPayPublicKeyFilePath);
  }

  public static class QueryOrderRequest {
    @SerializedName("transaction_id")
    @Expose(serialize = false)
    public String transactionId;
  
    @SerializedName("out_order_no")
    @Expose(serialize = false)
    public String outOrderNo;
  }
  
  public static class OrdersEntity {
    @SerializedName("transaction_id")
    public String transactionId;
  
    @SerializedName("out_order_no")
    public String outOrderNo;
  
    @SerializedName("order_id")
    public String orderId;
  
    @SerializedName("state")
    public OrderStatus state;
  
    @SerializedName("receivers")
    public List<OrderReceiverDetail> receivers = new ArrayList<OrderReceiverDetail>();
  }
  
  public enum OrderStatus {
    @SerializedName("PROCESSING")
    PROCESSING,
    @SerializedName("FINISHED")
    FINISHED
  }
  
  public static class OrderReceiverDetail {
    @SerializedName("amount")
    public Long amount;
  
    @SerializedName("description")
    public String description;
  
    @SerializedName("type")
    public ReceiverType type;
  
    @SerializedName("account")
    public String account;
  
    @SerializedName("result")
    public DetailStatus result;
  
    @SerializedName("fail_reason")
    public DetailFailReason failReason;
  
    @SerializedName("create_time")
    public String createTime;
  
    @SerializedName("finish_time")
    public String finishTime;
  
    @SerializedName("detail_id")
    public String detailId;
  }
  
  public enum ReceiverType {
    @SerializedName("MERCHANT_ID")
    MERCHANT_ID,
    @SerializedName("PERSONAL_OPENID")
    PERSONAL_OPENID
  }
  
  public enum DetailStatus {
    @SerializedName("PENDING")
    PENDING,
    @SerializedName("SUCCESS")
    SUCCESS,
    @SerializedName("CLOSED")
    CLOSED
  }
  
  public enum DetailFailReason {
    @SerializedName("ACCOUNT_ABNORMAL")
    ACCOUNT_ABNORMAL,
    @SerializedName("NO_RELATION")
    NO_RELATION,
    @SerializedName("RECEIVER_HIGH_RISK")
    RECEIVER_HIGH_RISK,
    @SerializedName("RECEIVER_REAL_NAME_NOT_VERIFIED")
    RECEIVER_REAL_NAME_NOT_VERIFIED,
    @SerializedName("NO_AUTH")
    NO_AUTH,
    @SerializedName("RECEIVER_RECEIPT_LIMIT")
    RECEIVER_RECEIPT_LIMIT,
    @SerializedName("PAYER_ACCOUNT_ABNORMAL")
    PAYER_ACCOUNT_ABNORMAL,
    @SerializedName("INVALID_REQUEST")
    INVALID_REQUEST
  }
  
}
